### PR TITLE
metamanager match pod messages for edged

### LINKED ISSUE #3489: cloud container be scheduled to the edge node
**What happened**:
this is my cluster info, it work on kubeedge in v1.8.2 version.
1 normal-first

#### Testing process #3293 
1. prepare
`systemctl stop edgecore` on all of the edgenode.
and clean all of the config on the edgenode
```shell
rm -f ca/*
rm -f certs/*
rm -fr edgemesh/
rm -f /var/lib/kubeedge/edgecore.db
```
and remove the `reliablesyncs.kubeedge.io` and it's children node from etcd
1 1 remove-reliablesyncs-kubeedge-io

and clean cloudimage on cloud node
`docker rmi kubeedge/cloudcore:v1.8.2`

2. modify these two files according to 3293
modify the `cloud/pkg/dynamiccontroller/application/listener.go` and `pkg/metaserver/util/util.go` 
In order to find the problem easier, I modified the `edged.go` file and added the print content:
```go
// klog.Infof("result content is %s", result.Content)
klog.Infof("result content is %s, resource: %s, group: %s, source: %s, operation: %s", result.Content, result.GetResource(), result.GetGroup(), result.GetSource(), result.GetOperation())
```
then rebuild the cloudimage and edgecore binary
```shell
make cloudimage
make all WHAT=edgecore
```
3. restart the cloudcore and add edgecore
2 add-edgenode
exec the `kubectl get pod -owide -nkubeedge -w` on the cloud

4. start the edgemesh-server and edgemesh-agent
At this moment, there is a problem that various cloud containers are scheduled to edge nodes.
It is found that containers such as `cloudcore` and `edgemesh` are also failing and restarting continuously;
3 watch-pod-status
The container being pulled up can be seen on the edge node: `docker ps -a`
4 docker-pa-a-on-edgenode

5. I keep the journal log files on the edge node
[edge-container-list.log](https://github.com/kubeedge/kubeedge/files/7749877/edge-container-list.log)
[edge-journal-3293.log](https://github.com/kubeedge/kubeedge/files/7749881/edge-journal-3293.log)

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:
I have tested a lot of this, and it has nothing to do with the number of edge nodes. The current positioning is the problem caused by fix #3293; because I don't know much about `synccontroller` and `dynamiccontroller`, I sent the log for answers;
**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
v1.17.17

- KubeEdge version(e.g. `cloudcore --version` and `edgecore --version`):
v1.8.2 + #3293 

- <details><summary>Cloud nodes Environment:</summary>

  - Hardware configuration (e.g. `lscpu`):
  - OS (e.g. `cat /etc/os-release`):
  - Kernel (e.g. `uname -a`):
  - Go version (e.g. `go version`):
  - Others:

  </details>
- <details><summary>Edge nodes Environment:</summary>

  - edgecore version (e.g. `edgecore --version`):
  - Hardware configuration (e.g. `lscpu`):
  - OS (e.g. `cat /etc/os-release`):
  - Kernel (e.g. `uname -a`):
  - Go version (e.g. `go version`):
  - Others:

  </details>

### PR DESCRIPTION
**Pull request by yz271544**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
In order to ensure that the function of edge's watch is not affected, add a function `podMatchByNodeNameOrHostIP` for metamanager::process to fix the issue of #3489 that the cloud application is scheduled to the edge;

**Which issue(s) this PR fixes**:

Fixes #3489

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: edge/pkg/metamanager/process.go [hunk 1]
```diff
@@ -132,7 +132,14 @@ func (m *metaManager) processInsert(message model.Message) {
...
 	// Notify edged
-	sendToEdged(&message, false)
+	switch resType {
+	case "pod":
+		if podMatchByNodeNameOrHostIP(&message, metaManagerConfig.Config.NodeName, metaManagerConfig.Config.NodeIP) {
+			sendToEdged(&message, false)
+		}
+	default:
+		sendToEdged(&message, false)
+	}
```
Comment (fisherxu) [inline edge/pkg/metamanager/process.go:142]: We can wrap a function to avoid duplication of code

File: edge/pkg/metamanager/process.go [hunk 2]
```diff
@@ -477,3 +491,41 @@ func (m *metaManager) runMetaManager() {
...
+	if status != nil {
+		if _, ok := status["hostIP"]; ok {
+			hostIPFromResource := status["hostIP"].(string)[0:]
+			isMatched = hostIPFromResource == hostIP
+			if isMatched {
+				return
+			}
+		}
+	}
+	return
```
Comment (fisherxu) [inline edge/pkg/metamanager/process.go:530]: We have the resourceType in the message, can identify pod using it. Besides, just nodename is enough? :)

File: edge/pkg/metamanager/process.go (hunk 1 above)
Comment (yz271544) [reply edge/pkg/metamanager/process.go]: yes, wrapped the a func sendToEdgedByNodeNameOrHostIP(message model.Message, resType string, isSync bool)

**Comment by yz271544** [comment]
At present, I really just judged `spec.nodeName` and `status.hostIP`. 

The reason is that I checked those pods specified by affinity through the log, and these two fields will be automatically added; 
whether it is enough to meet our matching conditions, we really need to continue Inquire about relevant documents and materials, or a large number of scene tests; 

In order to pass the integration test, I modified `helpers.go` and added a `nodeName` field. 

At present, this PR is issued, and I hope to get help from the community. 
**Please consider it carefully.**

**Comment by yz271544** [comment]
RFC ?

File: edge/pkg/metamanager/process.go [hunk 3]
```diff
@@ -175,7 +176,7 @@ func (m *metaManager) processUpdate(message model.Message) {
 		resp := message.NewRespByMessage(&message, OK)
 		sendToEdged(resp, message.IsSync())
 	case cloudmodules.EdgeControllerModuleName, cloudmodules.DynamicControllerModuleName:
-		sendToEdged(&message, message.IsSync())
```
Comment (fisherxu) [inline edge/pkg/metamanager/process.go:178]: Should not modify the `message.IsSync()`?

File: edge/pkg/metamanager/process.go [hunk 4]
```diff
@@ -477,3 +478,46 @@ func (m *metaManager) runMetaManager() {
 		}
 	}()
 }
+
+func sendToEdgedByNodeNameOrHostIP(message model.Message, resType string, isSync bool) {
+	switch resType {
+	case "pod":
```
Comment (fisherxu) [inline edge/pkg/metamanager/process.go:484]: Use the const same as in edged.

File: edge/pkg/metamanager/process.go [hunk 5]
```diff
@@ -477,3 +478,46 @@ func (m *metaManager) runMetaManager() {
...
+		if podMatchByNodeNameOrHostIP(message, metaManagerConfig.Config.NodeName) {
+			sendToEdged(&message, isSync)
+		}
+	default:
+		sendToEdged(&message, isSync)
+	}
+}
+
+func podMatchByNodeNameOrHostIP(message model.Message, nodeName string) (isMatched bool) {
+	content := message.GetContent()
```
Comment (fisherxu) [inline edge/pkg/metamanager/process.go:494]: We can json.Unmarshal  the content to pod directly.

File: edge/pkg/metamanager/process.go (hunk 3 above)
Comment (yz271544) [reply edge/pkg/metamanager/process.go:178]: yes, typo, corrected

File: edge/pkg/metamanager/process.go (hunk 4 above)
Comment (yz271544) [reply edge/pkg/metamanager/process.go]: OK, modified

File: edge/pkg/metamanager/process.go (hunk 5 above)
Comment (yz271544) [reply edge/pkg/metamanager/process.go:495]: OK, modified

**Pull request closed without merging**